.github/deployment/scripts/protocol-cli.rs

```diff
@@ -804,29 +872,64 @@ async fn main() -> Result<()> {
+                validate_addresses(&config, &format!("{}/{}/.env-otim-{}", env_dir, get_chain_info(chain).unwrap().network, get_chain_info(chain).unwrap().network), update).await?;
+            }
+            info(&format!("✓ Validated {} chains", chains.len()));
+            Ok(())
         }
 
-        Commands::DeployContracts { config_file, private_key } => {
-            let config = load_config(&config_file)?;
-            let addresses = deploy_contracts(&config, private_key.as_deref()).await?;
-
-            // Write addresses to file
-            let json_content = serde_json::to_string_pretty(&addresses)?;
-            std::fs::write("/output/deployed-addresses.json", json_content)?;
-            info("Contract addresses written to /output/deployed-addresses.json");
+        Commands::DeployContracts { config_file, env_dir, private_key } => {
+            let config = load_deploy_config(&config_file)?;
+            let chains = config.chains.as_ref().filter(|c| !c.is_empty()).ok_or_else(|| anyhow!("No chains in config"))?;
 
+            let mut all_addresses = HashMap::new();
+            for chain in chains {
+                info(&format!("Deploying to {}", chain));
+                if let Some(dir) = &env_dir {
+                    load_chain_env_files(dir, chain, get_chain_info(chain).ok_or_else(|| anyhow!("Unknown chain: {}", chain))?.network)?;
+                }
+                all_addresses.insert(chain.clone(), deploy_contracts(&config, private_key.as_deref()).await?);
+            }
+            std::fs::write("deployed-addresses.json", serde_json::to_string_pretty(&all_addresses)?)?;
+            info(&format!("✓ Deployed to {} chains → deployed-addresses.json", chains.len()));
             Ok(())
         }
 
-        Commands::WhitelistActions { addresses_file, private_key } => {
-            whitelist_actions(&addresses_file, private_key.as_deref()).await
+        Commands::WhitelistActions { config_file, addresses_file, env_dir, private_key } => {
+            let config = load_deploy_config(&config_file)?;
+            let chains = config.chains.as_ref().filter(|c| !c.is_empty()).ok_or_else(|| anyhow!("No chains in config"))?;
+            let addresses_by_chain: HashMap<String, HashMap<String, String>> =
+                serde_json::from_str(&fs::read_to_string(&addresses_file)?)?;
+
+            // Ensure all config chains have deployed addresses
+            let missing: Vec<_> = chains.iter().filter(|c| !addresses_by_chain.contains_key(*c)).collect();
+            if !missing.is_empty() { bail!("Topology mismatch: {:?} missing from addresses file", missing); }
+
+            // Warn about extra chains in addresses file
+            let extra: Vec<_> = addresses_by_chain.keys().filter(|k| !chains.contains(k)).collect();
+            if !extra.is_empty() { warn(&format!("Skipping chains not in config: {:?}", extra)); }
+
+            for chain in chains {
+                info(&format!("Whitelisting {}", chain));
```

Comment:
let's use tracing for this, it's going to be much easier to understand and more human readable. + you can use named arguments without having to interpolate them.